File: ServicesDashboard/Models/SystemMetrics.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Text.Json.Serialization;

namespace ServicesDashboard.Models;

/// <summary>
/// System-level metrics including network totals and temperatures
/// </summary>
public class SystemMetricsHistory
{
    [Key]
    public int Id { get; set; }

    [Required]
    public int ServerId { get; set; }

    public DateTime Timestamp { get; set; } = DateTime.UtcNow;

    // Network (aggregated across all interfaces)
    public long NetworkRxBytes { get; set; }
    public long NetworkTxBytes { get; set; }
    public long NetworkRxBytesPerSec { get; set; }
    public long NetworkTxBytesPerSec { get; set; }

    // Temperature (primary sensors)
    public float? CpuTemperature { get; set; }
    public float? GpuTemperature { get; set; }

    // Navigation property
    [JsonIgnore]
    [ForeignKey("ServerId")]
    public virtual ManagedServer Server { get; set; } = null!;
}

/// <summary>
/// Per-disk metrics for array, cache, and system disks
/// </summary>
public class DiskMetricsHistory
{
    [Key]
    public int Id { get; set; }

    [Required]
    public int ServerId { get; set; }

    public DateTime Timestamp { get; set; } = DateTime.UtcNow;

    // Disk identification
    [Required]
    [MaxLength(100)]
    public string DiskName { get; set; } = string.Empty;

    [Required]
    [MaxLength(50)]
    public string DiskType { get; set; } = string.Empty; // "array", "cache", "parity", "system"

    [MaxLength(255)]
    public string? MountPoint { get; set; }

    [MaxLength(50)]
    public string? Device { get; set; } // e.g., "/dev/sda"

    // Usage metrics
    public long TotalBytes { get; set; }
    public long UsedBytes { get; set; }
    public long FreeBytes { get; set; }
    public float UsagePercentage { get; set; }

    // Temperature (for drives)
    public float? Temperature { get; set; }

    // Status (for Unraid - active, standby, spun_down)
    [MaxLength(50)]
    public string? Status { get; set; }

    // Navigation property
    [JsonIgnore]
    [ForeignKey("ServerId")]
    public virtual ManagedServer Server { get; set; } = null!;
}

/// <summary>
/// Per-network-interface metrics
/// </summary>
public class NetworkInterfaceMetricsHistory
{
    [Key]
    public int Id { get; set; }

    [Required]
    public int ServerId { get; set; }

    public DateTime Timestamp { get; set; } = DateTime.UtcNow;

    // Interface identification
    [Required]
    [MaxLength(100)]
    public string InterfaceName { get; set; } = string.Empty;

    [Required]
    [MaxLength(50)]
    public string InterfaceType { get; set; } = string.Empty; // "physical", "bridge", "docker", "virtual"

    // Traffic (cumulative from boot)
    public long RxBytes { get; set; }
    public long TxBytes { get; set; }

    // Rate (calculated from delta between readings)
    public long RxBytesPerSec { get; set; }
    public long TxBytesPerSec { get; set; }

    // Navigation property
    [JsonIgnore]
    [ForeignKey("ServerId")]
    public virtual ManagedServer Server { get; set; } = null!;
}

File: services-dashboard-frontend/src/types/SystemMetrics.ts
// System-level metrics data point
export interface SystemMetricsDataPoint {
  timestamp: string;
  networkRxBytesPerSec: number;
  networkTxBytesPerSec: number;
  networkRxBytes: number;
  networkTxBytes: number;
  cpuTemperature: number | null;
  gpuTemperature: number | null;
}

// Temperature data point for charts
export interface TemperatureDataPoint {
  timestamp: string;
  cpuTemperature: number | null;
  gpuTemperature: number | null;
}

// Network interface data point
export interface NetworkInterfaceDataPoint {
  timestamp: string;
  rxBytesPerSec: number;
  txBytesPerSec: number;
}

// Network interface summary
export interface NetworkInterfaceSummary {
  interfaceName: string;
  interfaceType: 'physical' | 'bridge' | 'docker' | 'wireless' | 'virtual';
  currentRxBytesPerSec: number;
  currentTxBytesPerSec: number;
  avgRxBytesPerSec: number;
  avgTxBytesPerSec: number;
  maxRxBytesPerSec: number;
  maxTxBytesPerSec: number;
  history: NetworkInterfaceDataPoint[];
}

// Network metrics summary
export interface NetworkMetricsSummary {
  currentRxBytesPerSec: number;
  currentTxBytesPerSec: number;
  avgRxBytesPerSec: number;
  avgTxBytesPerSec: number;
  maxRxBytesPerSec: number;
  maxTxBytesPerSec: number;
  totalRxBytes: number;
  totalTxBytes: number;
  history: SystemMetricsDataPoint[];
}

// Temperature metrics summary
export interface TemperatureMetricsSummary {
  currentCpuTemperature: number | null;
  currentGpuTemperature: number | null;
  avgCpuTemperature: number | null;
  maxCpuTemperature: number | null;
  avgGpuTemperature: number | null;
  maxGpuTemperature: number | null;
  history: TemperatureDataPoint[];
}

// Server system metrics response
export interface ServerSystemMetricsResponse {
  serverId: number;
  serverName: string;
  serverType: 'unraid' | 'linux';
  lastUpdated: string;
  network: NetworkMetricsSummary;
  temperatures: TemperatureMetricsSummary;
  interfaces: NetworkInterfaceSummary[];
}

// Disk history data point
export interface DiskHistoryDataPoint {
  timestamp: string;
  usedBytes: number;
  usagePercentage: number;
  temperature: number | null;
}

// Disk information
export interface DiskInfo {
  diskName: string;
  diskType: 'array' | 'cache' | 'parity' | 'system' | 'data';
  device: string | null;
  mountPoint: string | null;
  totalBytes: number;
  usedBytes: number;
  freeBytes: number;
  usagePercentage: number;
  temperature: number | null;
  status: string | null;
  history: DiskHistoryDataPoint[];
}

// Disk metrics summary
export interface DiskMetricsSummary {
  // Unraid array
  arrayTotalBytes: number;
  arrayUsedBytes: number;
  arrayFreeBytes: number;
  arrayUsagePercentage: number;

  // Unraid cache
  cacheTotalBytes: number | null;
  cacheUsedBytes: number | null;
  cacheFreeBytes: number | null;
  cacheUsagePercentage: number | null;

  // Unraid parity
  parityTotalBytes: number | null;

  // Generic Linux
  systemTotalBytes: number;
  systemUsedBytes: number;
  systemFreeBytes: number;
  systemUsagePercentage: number;
}

// Server disk metrics response
export interface ServerDiskMetricsResponse {
  serverId: number;
  serverName: string;
  serverType: 'unraid' | 'linux';
  lastUpdated: string;
  summary: DiskMetricsSummary;
  disks: DiskInfo[];
}

// Chart data types for the frontend
export interface NetworkChartDataPoint {
  time: string;
  timestamp: number;
  rxBytesPerSec: number;
  txBytesPerSec: number;
}

export interface TemperatureChartDataPoint {
  time: string;
  timestamp: number;
  cpu: number | null;
  gpu: number | null;
}

export interface DiskChartDataPoint {
  name: string;
  type: string;
  total: number;
  used: number;
  free: number;
  percentage: number;
  temperature: number | null;
}
